### PR TITLE
doc(doc/config.md) : Documentation for OSM ConfigMap

### PR DESCRIPTION
**Pull request by snehachhabria**
**Description**:
 This PR adds documentation for OSM ConfigMap 

 Fixes : #1842 

**Affected area**:

- New Functionality      [ ]
- Documentation          [X]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution? `no`

### CODE REVIEW

File: docs/config.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# OSM ConfigMap
+
+OSM deploys a configMap `osm-config` as a part of its control plane (in the same namespace as that of the osm-controller pod) which can be updated by the mesh owner/operator at any time. The purpose of this configMap is to provide the mesh owner/operator the ability to reconfigure/update some of the mesh configurations based on their needs.
```
Comment (shashankram) [inline docs/config.md:3]:
```suggestion
OSM deploys a configMap `osm-config` as a part of its control plane (in the same namespace as that of the osm-controller pod) which can be updated by the mesh owner/operator at any time. The purpose of this configMap is to provide the mesh owner/operator the ability to update some of the mesh configurations based on their needs. 
```

File: docs/config.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+|-----|------|-----------------|---------------|----------|
+| permissive_traffic_policy_mode | bool | true, false | `"false"` | Setting to `true`, enables allow-all mode on the mesh i.e. all services in the mesh will be able to talk to one and other. If set to `false`, enables deny-all on mesh i.e. an `SMI Traffic Target` is necessary for a service in the mesh to talk to another service in the mesh. |
+| egress | bool | true, false| `"false"` | Enables egress on the mesh. |
+| enable_debug_server | bool | true, false| `"true"` | Enables the debug endpoint on the osm-controller pod to inspect and list most of the common structures used by the control plane at runtime. |
+| envoy_log_level | string | trace, debug, info, warning, warn, error, critical, off | `"error"` | Sets the verbosity of the logging of Envoy's joining the OSM service mesh. |
+| prometheus_scraping | bool | true, false | `"true"` | Enables a Prometheus listener on Envoy's which inturn enables Promethues metrics scraping in OSM. |
+| service_cert_validity_duration | string | 24h, 1m, 30s (any time duration) | `"24h"` | Sets the validity duration of mesh service certificates. |
```
Comment (shashankram) [inline docs/config.md:14]:
```suggestion
| service_cert_validity_duration | string | 24h, 1h30m (any time duration) | `"24h"` | Sets the service certificatevalidity duration, represented as a sequence of decimal numbers each with optional fraction and a unit suffix. |
```

File: docs/config.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# OSM ConfigMap
+
+OSM deploys a configMap `osm-config` as a part of its control plane (in the same namespace as that of the osm-controller pod) which can be updated by the mesh owner/operator at any time. The purpose of this configMap is to provide the mesh owner/operator the ability to reconfigure/update some of the mesh configurations based on their needs. 
+
+## ConfigMap Values
+
+| Key | Type | Possible Values | Default Value | Function |
```
Comment (shashankram) [inline docs/config.md:7]:
```suggestion
| Key | Type | Allowed Values | Default Value | Function |
```

File: docs/config.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+|-----|------|-----------------|---------------|----------|
+| permissive_traffic_policy_mode | bool | true, false | `"false"` | Setting to `true`, enables allow-all mode on the mesh i.e. all services in the mesh will be able to talk to one and other. If set to `false`, enables deny-all on mesh i.e. an `SMI Traffic Target` is necessary for a service in the mesh to talk to another service in the mesh. |
+| egress | bool | true, false| `"false"` | Enables egress on the mesh. |
```
Comment (shashankram) [inline docs/config.md:10]:
```suggestion
| egress | bool | true, false| `"false"` | Enables egress in the mesh. |
```

File: docs/config.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# OSM ConfigMap
+
+OSM deploys a configMap `osm-config` as a part of its control plane (in the same namespace as that of the osm-controller pod) which can be updated by the mesh owner/operator at any time. The purpose of this configMap is to provide the mesh owner/operator the ability to reconfigure/update some of the mesh configurations based on their needs. 
+
+## ConfigMap Values
+
+| Key | Type | Possible Values | Default Value | Function |
+|-----|------|-----------------|---------------|----------|
+| permissive_traffic_policy_mode | bool | true, false | `"false"` | Setting to `true`, enables allow-all mode on the mesh i.e. all services in the mesh will be able to talk to one and other. If set to `false`, enables deny-all on mesh i.e. an `SMI Traffic Target` is necessary for a service in the mesh to talk to another service in the mesh. |
```
Comment (shashankram) [inline docs/config.md:9]:
```suggestion
| permissive_traffic_policy_mode | bool | true, false | `"false"` | Setting to `true`, enables permissive mode in the mesh i.e. no traffic policy enforcement in the mesh. If set to `false`, enables deny-all traffic policy in mesh i.e. an `SMI Traffic Target` is necessary for services to communicate. |
```

File: docs/config.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+| egress | bool | true, false| `"false"` | Enables egress on the mesh. |
+| enable_debug_server | bool | true, false| `"true"` | Enables the debug endpoint on the osm-controller pod to inspect and list most of the common structures used by the control plane at runtime. |
```
Comment (shashankram) [inline docs/config.md:11]: What does `list most of the common structures` mean here?

File: docs/config.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+| envoy_log_level | string | trace, debug, info, warning, warn, error, critical, off | `"error"` | Sets the verbosity of the logging of Envoy's joining the OSM service mesh. |
```
Comment (shashankram) [inline docs/config.md:12]:
```suggestion
| envoy_log_level | string | trace, debug, info, warning, warn, error, critical, off | `"error"` | Sets the logging verbosity of Envoy proxy sidecar, only applicable to newly created pods joining the mesh. |
```

File: docs/config.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+| prometheus_scraping | bool | true, false | `"true"` | Enables a Prometheus listener on Envoy's which inturn enables Promethues metrics scraping in OSM. |
```
Comment (shashankram) [inline docs/config.md:13]:
```suggestion
| prometheus_scraping | bool | true, false | `"true"` | Enables Prometheus metrics scraping on sidecar proxies. |
```

**Review by shashankram — CHANGES_REQUESTED** [review]
This is a very useful change, left some comments regarding wording.

File: docs/config.md (hunk 6 above)
Comment (snehachhabria) [reply docs/config.md:11]: it refers to the structs/objects that OSM creates

Comment (shashankram) [reply docs/config.md:11]: I would change this to something users can comprehend: `Enables a debug endpoint on the osm-controller pod to list information regarding the mesh such as proxy connections, certificates, and SMI policies, .

File: docs/config.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# OSM ConfigMap
+
+OSM deploys a configMap `osm-config` as a part of its control plane (in the same namespace as that of the osm-controller pod) which can be updated by the mesh owner/operator at any time. The purpose of this configMap is to provide the mesh owner/operator the ability to update some of the mesh configurations based on their needs.
+
+## ConfigMap Values
+
+| Key | Type | Allowed Values | Default Value | Function |
+|-----|------|-----------------|---------------|----------|
+| permissive_traffic_policy_mode | bool | true, false | `"false"` | Setting to `true`, enables allow-all mode in the mesh i.e. no traffic policy enforcement in the mesh. If set to `false`, enables deny-all traffic policy in mesh i.e. an `SMI Traffic Target` is necessary for services to communicate. |
```
Comment (shashankram) [inline docs/config.md:9]:
```suggestion
| permissive_traffic_policy_mode | bool | true, false | `"false"` | Setting to `true`, enables permissive mode, i.e. no traffic policy enforcement in the mesh. If set to `false`, enables deny-all traffic policy in mesh i.e. an `SMI Traffic Target` is necessary for services to communicate. |
```

**Review by shashankram — APPROVED** [review]
Thank you!

**Comment by shashankram** [comment]
@snehachhabria should we rename the doc to `docs/osm_config_map.md`?

**Comment by snehachhabria** [comment]
> @snehachhabria should we rename the doc to `docs/osm_config_map.md`?

we could, would that be better ?

**Review by eduser25 — APPROVED** [review]

**Review by shashankram — APPROVED** [review]

**Review by eduser25 — APPROVED** [review]

**Pull request merged**